ocaml/xapi/repository.ml

```diff
@@ -583,3 +702,328 @@ let get_repository_handler (req : Http.Request.t) s _ =
+          in
+          debug "updateinfo: %s" (string_of_updateinfo ui);
+          Some ui
+        | _ -> None
+    ) children
+    |> List.map (fun updateinfo -> (updateinfo.id, updateinfo))
+  | _ ->
+    error "Failed to parse updateinfo.xml: missing <updates>";
+    raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
+  | exception e ->
+    error "Failed to parse updateinfo.xml: %s" (ExnHelper.string_of_exn e);
+    raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
+
+let parse_updateinfo ~hash =
+  let repo_dir = Filename.concat !Xapi_globs.local_pool_repo_dir !Xapi_globs.pool_repo_name in
+  let repodata_dir = Filename.concat repo_dir "repodata" in
+  let repomd_xml_path = Filename.concat repodata_dir "repomd.xml" in
+  let md = parse_repomd repomd_xml_path in
+  if hash <> md.checksum then
+    (error "Unexpected mismatch between XAPI DB and YUM DB. Need to do pool.sync-updates again.";
+     raise Api_errors.(Server_error (createrepo_failed, [])));
+  let updateinfo_xml_gz_path = Filename.concat repo_dir md.location in
+  match Sys.file_exists updateinfo_xml_gz_path with
+  | false ->
+    error "File %s doesn't exist" updateinfo_xml_gz_path;
+    raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
+  | true ->
+    with_updateinfo_xml updateinfo_xml_gz_path parse_updateinfo_xml_file
+
+let compare_version_strings s1 s2 =
+  (* Compare versions or releases of RPM packages
+   * I.E. for "libpath-utils-0.2.1-29.el7.x86_64" and "libpath-utils-0.2.1a-30.el7.x86_64",
+   * this function compares:
+   * versions between "0.2.1" and "0.2.1a", or
+   * releases between "29.el7" and "30.el7".
+   * More examples:
+   *  "1.2.3" "<" "1.2.4"
+   *  "1.2.3" "=" "1.2.3"
+   *  "1.2.3" ">" "1.2"
+   *  "1.0011" ">" "1.9"
+   *  "1.05" "=" "1.5"
+   *  "1.0" ">" "1"
+   *  "1.0" ">" "1.a"
+   *  "2.50" ">" "2.5"
+   *  "XS3" "<" "xs2"
+   *  "1.2.3" ">" "1.2.3a"
+   *  "xs4" "=" "xs.4"
+   *  "2a" "<" "2.0"
+   *  "2a" "<" "2b"
+   *  "1.0" ">" "1.xs2"
+   *  "1.0_xs" "=" "1.0.xs"
+   *)
+  let normalize v =
+    let split_letters_and_numbers s =
+      let r = Re.Str.regexp "^\\([^0-9]+\\)\\([0-9]+\\)$" in
+      if Re.Str.string_match r s 0 then
+        [(Re.Str.replace_first r "\\1" s); (Re.Str.replace_first r "\\2" s)]
+      else [s]
+    in
+    v
+    |> Astring.String.cuts ~sep:"."
+    |> List.map (fun s -> Astring.String.cuts ~sep:"_" s)
+    |> List.flatten
+    |> List.map (fun s -> split_letters_and_numbers s)
+    |> List.flatten
+    |> List.map (fun s -> try Int (int_of_string s) with _ -> Str s)
+  in
+  let compare_str s1 s2 f =
+    let r = String.compare s1 s2 in
+    if r < 0 then Lt
+    else if r > 0 then Gt
+    else f ()
+  in
+  let rec compare_segments l1 l2 =
+    match l1, l2 with
+    | c1 :: t1, c2 :: t2 ->
+      let f () = compare_segments t1 t2 in
+      (match c1, c2 with
+       | Int s1, Int s2 ->
+         if s1 > s2 then Gt
+         else if s1 = s2 then f ()
```

Comment:
If you inline f, then you can get the compiler to check that your recursive calls are tail recursive with `(compare_segments [@tailcall]) t1 t2`

Reply:
Thanks. It's great to learn about this annotation. :)
I resolved the `f` completely as I found no more usages on it.